### ISSUE TITLE
Firestore Document annotations are confusing (fail) when used together

### ISSUE THREAD

**Issue by SteveBurkert** [firebase/firebase-android-sdk#6601]
### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: Android Studio Ladybug | 2024.2.1 Patch 2
  * Firebase Component: Firestore
  * Component version: BOM 33.6.0

### [REQUIRED] Step 3: Describe the problem

Discussed in https://github.com/firebase/firebase-android-sdk/discussions/6600

<div type='discussions-op-text'>

<sup>Originally posted by **SteveBurkert** December 13, 2024</sup>
Before I start a wall of text explaining myself, I try to explain my issue with an example.

Let's say we have a very basic kotlin data class, that we want to be serialized as a document
to the firestore. It could look like this:

```kotlin
data class MyDocument(
    val myFoo: String
)
```

Let's say we initialize this with "hey". 
We know this class will end up being parsed by the processor / mapper: [CustomClassMapper](https://github.com/firebase/firebase-android-sdk/blob/main/firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java)

And in firestore we will see a beautiful entry holding "myFoo" = "hey" under some document id.

But now let's say we want to change the name of the serialized field, but still want to use camelCase in the kotlin code. In firestore world we might want underscores. So we **want to use different property names when serialized**.
Like this: 

```kotlin
data class MyDocument(
    @PropertyName("my_foo")
    val myFoo: String
)
```

Funny thing, this won't work. 
It will still be serialized as "myFoo". 

**The reason this does not work**

Kotlin creates a public getter and a private backing field for myFoo.
While parsing the mapper will see a public getter, that has **no** annotation and store a value inside his list of properties -> "myFoo" or "myfoo".

The mapper then will see the private field and also see the fields `PropertyName` annotation, resolving the name to "my_foo", which IS NOT inside the properties list and hence, the field and also all of it's annotations (PropertyName, ServerTimestamp, DocumentId) are IGNORED.

<img width="556" alt="Screenshot 2024-12-13 at 13 30 36" src="https://github.com/user-attachments/assets/b24ee5ec-f5c9-4cd0-84bb-2a148435c2f6" />

So basically, the @PropertyName annotation without use-site target is ignored IF the value
is different to the fields name. (Which is the purpose of the annotation? Right?)

But this I can understand. If we would not ignore it, we would end up storing 2 
properties / fields for actually 1 field and have then 2 entries in our document in the cloud.

**First slight request.**

Even though I am not entirely sure how to handle this, I still think this could be done differently. Sure we should not store 1 property and 1 field, but we could still honour the annotations of the field and the getter, by mapping between getter/field name and PropertyName.value.

--- 

**But now we come to the part that I found really confusing:** 

In order to resolve our issue we might use use-site targets like so: 

```kotlin
data class MyDocument(
    @get:PropertyName("my_foo")
    val myFoo: String
)
```

This *will* work. The getter is used for serializing and the backing field...is ignored.

Is ignored? Yes - and all of it's annotations, too.

So let's assume I want to store a serverside timestamp inside my foo. 

```kotlin
data class MyDocument(
    @get:PropertyName("my_foo")
    @ServertimeStamp
    val myFoo: Timestamp? = null
)
```

This will **not** work as expected and there will be **no** timestamp generated on the server and the value will be `null`.

`@ServertimeStamp` will work without PropertyName, or with the field having the same name as the value of PropertyName. 

Now we have an annotation on the getter and one annotation on the field. 
Because of refactoring, change, whatever this might happen.
For example, if your code was previously 

```kotlin
data class MyDocument(
    @get:PropertyName("myFoo")
    @ServertimeStamp
    val myFoo: Timestamp? = null
)
```

all is good. But now you changed it to: 

```kotlin
data class MyDocument(
    @get:PropertyName("my_foo")
    @ServertimeStamp
    val myFoo: Timestamp? = null
)
```

And `my_foo` ends up in firestore...but huh? No timestamp and the value is `null`. 

**Conclusion**

I think this is a counterintuitive behaviour that must at least be documented somewhere OR 
field site annotations **should still be honoured**, even if the resolved field name is not already
in the list of properties.

There are plenty of stackoverflow questions regarding the `@PropertyName` annotation not working
or weird serialization that all have to do with the same issue. 

Also, even though documented, it feels weird that if you use `@PropertyName` within a document class,
you can use it for serialization, but you can not use the same class for deserialization.
Maybe there should at least be a runtime warning log. (There none I am aware of) </div>

**Comment by dconeybe** [comment]
@SteveBurkert You are right that this behavior is confusing. Obviously, the `CustomClassMapper` was designed for Java, and was retrofitted to support Kotlin. That being said, this could have been done better.

With the existing `CustomClassMapper` the general rule-of-thumb is to _always_ explicitly specify the use-site targets `@get` and/or `@set` (https://kotlinlang.org/docs/annotations.html#annotation-use-site-targets).

If the data class is _only_ used to _writing_ to Firestore, you can use a class like this:

```kt
data class MyDocument(
  @get:[PropertyName("my_foo") ServerTimestamp]
  val myFoo: Timestamp? // optionally, =null can be specified as the default value
)
```

To use the class for _both_ reading _and_ writing, you have to make a few changes:
1. Change `val` to `var` because CustomClassMapper requires that the field be mutable.
2. Add an `@set:PropertyName` annotation that matches the `@get:PropertyName` annotation.
3. Add an explicit no-arg constructor because CustomClassMapper requires one (or annotate the constructor with `@JvmOverloads` and provide default values for all parameters; see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.jvm/-jvm-overloads/).

```kt
data class MyDocument(
  @get:[PropertyName("my_foo") ServerTimestamp]
  @set:PropertyName("my_foo")
  var myFoo: Timestamp? // optionally, =null can be specified as the default value, but that can be confusing because, then, Kotlin will prefer the primary constructor over the no-arg constructor.
) {
  constructor() : this(null)
}
```

**Comment by dconeybe** [comment]
For posterity, I created a sample app to experiment with the various use-site target permutations: https://github.com/dconeybe/Issue6601KotlinDataClassCustomClassMapperAndroidTest/blob/main/app/src/androidTest/java/com/example/fstkotlin/DemoAndroidTest.kt

**Comment by dconeybe** [comment]
My personal opinion is that rather than further patching CustomClassMapper that Firestore should add support for `kotlinx.serialization` (https://kotlinlang.org/docs/serialization.html). Then, you could just do something like this:

```kt
@file:UseSerializers(TimestampSerializer::class)

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.PropertyName
import com.google.firebase.firestore.ServerTimestamp
import com.google.firebase.firestore.TimestampSerializer
import com.google.firebase.firestore.decode
import com.google.firebase.firestore.set as setSerialized
import com.google.firebase.firestore.set
import kotlinx.coroutines.tasks.await
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.serializer

@Serializable
data class MyDocument(
  @PropertyName("my_foo")
  @ServerTimestamp
  val myFoo: Timestamp? = null
)

suspend fun demo(db: FirebaseFirestore) {
  val documentReference = db.document("foo/bar")
  documentReference.setSerialized(MyDocument()).await()
  documentReference.set(MyDocument(), serializer()).await() // alternative to the line above
  val data: MyDocument = documentReference.get().await().decode<MyDocument>()
  println("Server timestamp is: ${data.myFoo}")
}
```

Then Firestore would provide some extension functions like these:

```kt
package com.google.firebase.firestore

import com.google.android.gms.tasks.Task
import com.google.firebase.Timestamp
import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerializationStrategy
import kotlinx.serialization.UseSerializers
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.serializer

inline fun <reified T> DocumentSnapshot.decode(): T = decode(serializer())

fun <T> DocumentSnapshot.decode(deserializationStrategy: DeserializationStrategy<T>): T = TODO()


inline fun <reified T> DocumentReference.set(data: T): Task<Void> = set(data, serializer())

inline fun <reified T> DocumentReference.set(data: T, options: SetOptions): Task<Void> =
  set(data, options, serializer())


fun <T> DocumentReference.set(
  data: T,
  serializationStrategy: SerializationStrategy<T>
): Task<Void> = TODO()

fun <T> DocumentReference.set(
  data: T,
  options: SetOptions,
  serializationStrategy: SerializationStrategy<T>
): Task<Void> = TODO()


object TimestampSerializer : KSerializer<Timestamp> {
  ...
}
```

This is very close to what Firebase Data Connect does, by implementing `kotlinx.serialization.encoding.Encoder` and `kotlinx.serialization.encoding.Decoder` in https://github.com/firebase/firebase-android-sdk/blob/main/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/util/ProtoStructEncoder.kt and https://github.com/firebase/firebase-android-sdk/blob/main/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/util/ProtoStructDecoder.kt, respectively.

**Comment by dconeybe** [comment]
As it turns out, a lot of the work of integrating with kotlinx.serialization was done in https://github.com/firebase/firebase-android-sdk/pull/3951 but we never had a chance to fully code review and merge it.  Google employees see b/388841250 for more details.
